katcp/testutils.py

```diff
@@ -2049,3 +2051,41 @@ def _advance(self, future):
         yield self.test_instance.set_ioloop_time(
             self.test_instance.ioloop_time + self.quantum)
         future.set_result(None)
+
+
+@contextmanager
+def assert_no_memory_leaks():
+    """Context handler that asserts no memory is leaking within.
+
+    Example
+    -------
+
+    Creating object, then releasing reference:
+
+        >>> with assert_no_memory_leaks():
+        >>>     my_obj = MyClass()
+        >>>     my_obj.optionally_do_stuff()
+        >>>     my_obj = None
+
+    Raises
+    ------
+    AssertionError
+        If the garbage collector finds any unreachable objects when
+        exiting the context handler.
+
+    """
+    old_config = gc.get_debug()
+    gc.set_debug(gc.DEBUG_SAVEALL)
```

Comment:
Are you sure you want this? It means that the test will fail if there happens to be cyclic garbage left over that *can* be garbage-collected (just not as quickly as the reference-counter would free it), as opposed to only failing on uncollectable garbage.

If this is what you want, I would suggest:
- Setting the new flags to `old_config | gc.DEBUG_SAVEALL` so that any other flags set for debugging aren't wiped.
- Doing the initial `gc.collect()` before setting the flag, so that cyclic but collectable garbage that existed before the test can be freed rather than being stuck forever in `gc.garbage`.